src/libasr/pass/intrinsic_function_registry.h

```diff
@@ -1358,6 +1360,108 @@ namespace Sign {
+            return make_ConstantWithType(make_RealConstant_t, fval, return_type, loc);
+        }
+    }
+
+    static inline ASR::asr_t* create_Aint(
+        Allocator& al, const Location& loc, Vec<ASR::expr_t*>& args,
+        const std::function<void (const std::string &, const Location &)> err) {
+        Vec<ASR::expr_t *> m_args; m_args.reserve(al, 1);
+        m_args.push_back(al, args[0]);
+        if ( args[1] ) {
+            m_args.push_back(al, args[1]);
+            if (!ASR::is_a<ASR::IntegerConstant_t>(*m_args[1])) {
+                err("`kind` argument of the `aint` function must be an Integer",
+                    m_args[1]->base.loc);
+            }
+        }
+        if (!(m_args.size() == 1 || m_args.size() == 2)) {
+            err("Intrinsic `aint` function accepts exactly 1 or 2 arguments", loc);
+        }
+        ASR::ttype_t *type = ASRUtils::expr_type(m_args[0]);
+        if (!ASRUtils::is_real(*type)) {
+            err("Argument of the `aint` function must be Real",
+                m_args[0]->base.loc);
+        }
+        ASR::expr_t *m_value = nullptr;
+        if (all_args_evaluated(m_args)) {
+            Vec<ASR::expr_t*> arg_values; arg_values.reserve(al, 1);
+            arg_values.push_back(al, expr_value(m_args[0]));
+            if (m_args.size() == 2) {
+                arg_values.push_back(al, expr_value(m_args[1]));
+            }
+            m_value = eval_Aint(al, loc, expr_type(m_args[0]), arg_values);
+        }
+        return ASR::make_IntrinsicScalarFunction_t(al, loc,
+            static_cast<int64_t>(IntrinsicScalarFunctions::Aint),
+            m_args.p, m_args.n, 0, ASRUtils::expr_type(m_args[0]), m_value);
+    }
+
+    static inline ASR::expr_t* instantiate_Aint(Allocator &al, const Location &loc,
+        SymbolTable *scope, Vec<ASR::ttype_t*>& arg_types, ASR::ttype_t *return_type,
+        Vec<ASR::call_arg_t>& new_args, int64_t /*overload_id*/) {
+        std::string func_name = "_lcompilers_aint_" + type_to_str_python(arg_types[0]);
+        std::string fn_name = scope->get_unique_name(func_name);
+        SymbolTable *fn_symtab = al.make_new<SymbolTable>(scope);
+        Vec<ASR::expr_t*> args;
+        args.reserve(al, new_args.size());
+        ASRBuilder b(al, loc);
+        Vec<ASR::stmt_t*> body; body.reserve(al, 1);
+        SetChar dep; dep.reserve(al, 1);
+        if (scope->get_symbol(fn_name)) {
+            ASR::symbol_t *s = scope->get_symbol(fn_name);
+            ASR::Function_t *f = ASR::down_cast<ASR::Function_t>(s);
+            return b.Call(s, new_args, expr_type(f->m_return_var), nullptr);
+        }
+        fill_func_arg("a", arg_types[0]);
+        if (new_args.size() == 2) {
+            fill_func_arg("kind", arg_types[1]);
+        }
+
+        auto result = declare(fn_name, return_type, ReturnVar);
+
+        // Cast: Real -> Integer -> Real
+        // TODO: Use the kind argument
+        body.push_back(al, b.Assignment(result, i2r32(r2i64(args[0]))));
```

Comment:
```suggestion
        body.push_back(al, b.Assignment(result, i2r(r2i(args[0], int_type), arg_type)));
```
Compute int_type based on the arg_types, I mean int type with the same kind as arg_type

Reply:
int_type can be anything we wish, I feel let's be more accurate and use `r2i64`, for arg_type we will need `kind`, here `kind` can be fetched from `return_type`, we can set `return_type` in `create_Aint`

Reply:
Does this work for very large number, that doesn't even store i64?

Reply:
I feel we will come across this case even when we do `r2i` that will also have a bottle neck of 2^64 - 1, right?

Reply:
Yes yes, I just got that thought so shared.

Reply:
As we are sure that we need to return real*4 by default, lets do `i2r32` and put a TODO to handle kind, by utilizing arg_type we are doing the same stuff, I think